### PR TITLE
Add deprecation warning for OpenMP [openmp-warning-dev]

### PR DESCRIPTION
**Pull request by jdahm**

### CODE REVIEW

**Comment by tzanio** [comment]
Thanks Johann, this looks good.

I do have to suggestions though that I'd like to hear @v-dobrev's and your opinion on:

1) List the deprecated option in the warning message, e.g.
```c++
  #pragma message "[MFEM_USE_OPENMP]: " DEPRECATION_WARNING
```

2) Move the deprecated section in its own file, e.g. `config/deprecated.hpp` and include it only in one *.cpp file that is always compiled, e.g. `general/array.hpp` to avoid multiple copies of the warning message.

What do you think?

**Comment by jdahm** [comment]
Thanks for looking at this @tzanio. I agree with you; I don't like that it prints on every compilation unit. Another option, better in my opinion, is to add a PHONY target to the makefile and have the mfem library target depend on it. I just pushed a commit that adds this. We can revert commits from the version we choose not to use. Let's let @v-dobrev weigh in as well.

**Comment by tzanio** [comment]
I like this more, thanks @jdahm

File: config/config.hpp [hunk 1]
```diff
@@ -52,3 +52,11 @@
...
 #endif
 #endif // MFEM_USE_MPI not defined
+
+// Deprecated features:
+
+#define DEPRECATION_WARNING "This feature is planned for removal. Please let the MFEM team know if you rely on it."
+
+#ifdef MFEM_USE_OPENMP
+#pragma message DEPRECATION_WARNING
+#endif
```
Comment (tzanio) [inline config/config.hpp]: No longer necessary with the makefile approach?

Comment (jdahm) [reply config/config.hpp]: 👍

**Comment by tzanio** [comment]
@v-dobrev: can you take a quick look?

**Comment by v-dobrev** [comment]
As it is now, the message is easy to miss. It will be better to print that at the end of the build, separating it with some empty lines; adding some color will also make it easier to see.

**Comment by tzanio** [comment]
In light of https://github.com/mfem/mfem/pull/631#issuecomment-479229365, can we merge this, even if not optimal, so people don't get confused?

**Comment by jdahm** [comment]
Prints a red/yellow colored error after the build. I did rename the option to `MFEM_USE_LEGACY_OPENMP` for convenience to use some of the existing pattern matching in the makefile.

Test on my mac:
```
$ make config MFEM_CXX=g++-8 MFEM_USE_LEGACY_OPENMP=YES MFEM_THREAD_SAFE=YES
$ make
...
a - fem/tmop.o
ranlib -no_warning_for_no_symbols libmfem.a
[MFEM_USE_LEGACY_OPENMP]: This feature is planned for removal in the next release. Please open an issue at github.com/mfem/mfem/issues if you depend on it.```

**Comment by jdahm** [comment]
By that do you mean that you’d like to preserve the MFEM_USE_OPENMP option (just add the legacy option) in this PR so that there won’t be a merge conflict with Okina?

**Comment by tzanio** [comment]
> By that do you mean that you’d like to preserve the MFEM_USE_OPENMP option (just add the legacy option) in this PR so that there won’t be a merge conflict with Okina?

Yes, I was thinking that there should be two OpenMP options: `MFEM_USE_LEGACY_OPENMP` for the old behavior and `MFEM_USE_OPENMP` for use with Okina.

File: makefile [hunk 2]
```diff
@@ -231,8 +231,19 @@ ifeq ($(MFEM_USE_$(1)),YES)
...
+define mfem_add_legacy_dependency
+ifeq ($(MFEM_USE_LEGACY_$(1)),YES)
+   INCFLAGS += $($(1)_OPT)
+   ALL_LIBS += $($(1)_LIB)
+endif
+endef
+
 # Process dependencies
-$(foreach dep,$(MFEM_DEPENDENCIES),$(eval $(call mfem_add_dependency,$(dep))))
+$(foreach dep,$(MFEM_LEGACY_DEPENDENCIES),$(eval $(call mfem_add_legacy_dependency,$(dep))))
```
Comment (v-dobrev) [inline makefile]: This line should not replace the one with `$(MFEM_DEPENDENCIES)` -- both need to be there. This is seems to be the reason the travis tests failed.

**Comment by v-dobrev** [comment]
> > By that do you mean that you’d like to preserve the MFEM_USE_OPENMP option (just add the legacy option) in this PR so that there won’t be a merge conflict with Okina?
> 
> Yes, I was thinking that there should be two OpenMP options: `MFEM_USE_LEGACY_OPENMP` for the old behavior and `MFEM_USE_OPENMP` for use with Okina.

Aren't they the same option? E.g. `-fopenmp`.

File: makefile (hunk 2 above)
Comment (jdahm) [reply makefile]: Oh I meant to have both. That’s an error I’ll fix

**Comment by jdahm** [comment]
Yes, they’re adding the same compilation/link flag. That was my intention behind the mfem_add_legacy_dependency macro: so we wouldn’t have to write the option twice in defaults.mk

**Comment by jdahm** [comment]
> `MFEM_USE_OPENMP` should now provide the OpenMP compile flags, so we can use it in the okina branch

I added `MFEM_USE_OPENMP` back in the code again.

File: config/cmake/MFEMConfig.cmake.in [hunk 3]
```diff
@@ -24,7 +24,7 @@ set(MFEM_USE_GZSTREAM @MFEM_USE_GZSTREAM@)
 set(MFEM_USE_LIBUNWIND @MFEM_USE_LIBUNWIND@)
 set(MFEM_USE_LAPACK @MFEM_USE_LAPACK@)
 set(MFEM_THREAD_SAFE @MFEM_THREAD_SAFE@)
-set(MFEM_USE_OPENMP @MFEM_USE_OPENMP@)
+set(MFEM_USE_LEGACY_OPENMP @MFEM_USE_LEGACY_OPENMP@)
```
Comment (v-dobrev) [inline config/cmake/MFEMConfig.cmake.in]: We should have both options, right?

File: config/cmake/config.hpp.in [hunk 4]
```diff
@@ -63,7 +63,7 @@
 #cmakedefine MFEM_THREAD_SAFE
 
 // Enable experimental OpenMP support. Requires MFEM_THREAD_SAFE.
-#cmakedefine MFEM_USE_OPENMP
+#cmakedefine MFEM_USE_LEGACY_OPENMP
```
Comment (v-dobrev) [inline config/cmake/config.hpp.in]: Same here: we should have both options.

File: config/cmake/modules/MfemCmakeUtilities.cmake [hunk 5]
```diff
@@ -714,7 +714,7 @@ function(mfem_export_mk_files)
   # Convert Boolean vars to YES/NO without writting the values to cache
   set(CONFIG_MK_BOOL_VARS MFEM_USE_MPI MFEM_USE_METIS MFEM_USE_METIS_5
       MFEM_DEBUG MFEM_USE_EXCEPTIONS MFEM_USE_GZSTREAM MFEM_USE_LIBUNWIND
-      MFEM_USE_LAPACK MFEM_THREAD_SAFE MFEM_USE_OPENMP MFEM_USE_MEMALLOC
+      MFEM_USE_LAPACK MFEM_THREAD_SAFE MFEM_USE_LEGACY_OPENMP MFEM_USE_MEMALLOC
```
Comment (v-dobrev) [inline config/cmake/modules/MfemCmakeUtilities.cmake]: Same here: keep both options.

File: config/defaults.cmake [hunk 6]
```diff
@@ -26,7 +26,7 @@ option(MFEM_USE_GZSTREAM "Enable gzstream for compressed data streams." OFF)
 option(MFEM_USE_LIBUNWIND "Enable backtrace for errors." OFF)
 option(MFEM_USE_LAPACK "Enable LAPACK usage" OFF)
 option(MFEM_THREAD_SAFE "Enable thread safety" OFF)
-option(MFEM_USE_OPENMP "Enable OpenMP usage" OFF)
+option(MFEM_USE_LEGACY_OPENMP "Enable OpenMP usage" OFF)
```
Comment (v-dobrev) [inline config/defaults.cmake]: Same here: keep both options.

File: config/cmake/MFEMConfig.cmake.in (hunk 3 above)
Comment (jdahm) [reply config/cmake/MFEMConfig.cmake.in]: Good catch!

**Comment by v-dobrev** [comment]
@jdahm,

Do you have any objection to merging this in the `okina` branch instead of `master`? We can use it there to introduce the OpenMP backend.

@tzanio and @camierjs are on board with that.

**Comment by jdahm** [comment]
The GitHub online merge tool added some extra undesired changes to this PR when switching the merge target to okina. I need to remove these quickly before this can be merged.

**Comment by jdahm** [comment]
This should be ready to squash-merge onto okina now.

**Pull request merged**